### ISSUE TITLE
drive: Queries per 100 seconds per user

### ISSUE THREAD

**Issue by azgul** [ncw/rclone#1390]
Just a question, what is it for the rclone project? It defaults to 1,000 requests/100s per user, so 10 requests per sec.

Currently `minSleep` in the drive implementation is set to `10 ms` which implies that the quota for users is 10,000/100s (10x as many as a fresh project has). Is that correct or could it be why so many people are hit with soft bans for exceeding user quota?

**Comment by azgul** [comment]
So if this theory is correct it would be trivial to increase `minSleep` to `100ms` and never get rate limited on a user basis. I'll be testing this over the weekend.

**Comment by jaketame** [comment]
So it could be that @ncw has requested a rate increase for QPS hence the lower minSleep = 10 but for people using there own client id then this is too high and could be why people are getting API bans rather than download bans on syncing.

I've requested a rate increase on my own ID to see if it gets applied as this could be an alternate solution.

We could introduce a check to see if rclone config is using own client id and then change to the higher minSleep = 100

**Comment by calisro** [comment]
If I look at my project ID quotas.  For the past 30 days the highest I had was 365 queries per 100 seconds.

**Comment by jaketame** [comment]
I have been increased to 100QPS.

```
i've increased your per-user limit to 100 to match your global QPS. This will help with read operations. Kindly update your user-limit to the update value to make sure it reflects on your project.

There is also a hard limit for write operations (create, update, delete) that is lower than your allowed QPS. And the hard limit for operations is something that cannot be lifted. If you haven't done so already, please consider implementing exponential backoff for error retries.

Queries per 100 seconds per user	10,000	
Queries per 100 seconds	10,000	
```
I've not no errors based on that.

rlcone size gcd-crypt: --checkers=40 -v
Total objects: 7414
Total size: 16645.706 GBytes (17873190532742 Bytes)
2017/05/05 14:54:32 Go routines at exit 202

real	0m23.627s
user	0m2.016s
sys	0m0.424s

3 low level retry errors
rlcone size gcd-crypt: --checkers=99 -v
Total objects: 7414
Total size: 16645.706 GBytes (17873190532742 Bytes)
2017/05/05 14:57:40 Go routines at exit 202

real	0m21.005s
user	0m2.008s
sys	0m0.504s

**Comment by calisro** [comment]
You mind telling me what you told them to get it increased?  I'd like to test it.

**Comment by ncw** [comment]
@azgul  wrote:
> Just a question, what is it for the rclone project? It defaults to 1,000 requests/100s per user, so 10 requests per sec.

That is a long term average, but yes that is correct.

Rclone is set to `Queries per 100 seconds per user	1,000` so 10 per second as you said.

> Currently minSleep in the drive implementation is set to 10 ms which implies that the quota for users is 10,000/100s (10x as many as a fresh project has). Is that correct or could it be why so many people are hit with soft bans for exceeding user quota?

It is a possibility.  I set the quotas based on people doing file transfers which use less transactions per second.  The theory being you need a burst during the directory comparisons then not a lot during the actual file transfers.

> So if this theory is correct it would be trivial to increase minSleep to 100ms and never get rate limited on a user basis. I'll be testing this over the weekend.

Worth a try!

@jaketame wrote
> So it could be that @ncw has requested a rate increase for QPS hence the lower minSleep = 10 but for people using there own client id then this is too high and could be why people are getting API bans rather than download bans on syncing.

rclone has 1200 QPS now I think.  It is quite easy to get them to raise the limit to 100 QPS, above that it harder!

**Comment by azgul** [comment]
Tested it with minSleep `100ms` and I got download quota ban after around 12h of having it mounted. I guess due to the uncached nature of rclone that hits the drive API in a manner that wouldn't happen if it was cached. Since it wasn't visible in the API dashboard, it must be because whatever accessed the files must have used the download url to get file information.

So in other words: didn't get API quota ban, but download quota ban.
